pytorch_lightning/utilities/auto_restart.py

```diff
@@ -734,13 +742,95 @@ def _patch_dataloader_get_iterators() -> None:
+    # cleanup the get_iterator replacement in case of Fault-tolerance.
     get_iterator = getattr(DataLoader, "_ori_get_iterator", None)
     if get_iterator:
         DataLoader._get_iterator = get_iterator
         del DataLoader._ori_get_iterator
 
 
+def _validate_iterable_dataset(dataloader: DataLoader):
+    SUPPORTED_SAMPLERS = (RandomSampler, SequentialSampler, DistributedSampler)
+
+    dataset = dataloader.dataset
+
+    if getattr(dataset, "__next__", None) is None:
+        raise AttributeError(
+            "Fault-tolerance doesn't support an `IterableDataset` without `__next__` "
+            "method implemented. Hint: We recommend you to move your logic from `__iter__`"
+            " inside and rely on a sampler to perform the sample sampling."
+        )
+
+    samplers = {k: v for k, v in dataset.__dict__.items() if isinstance(v, Sampler)}
+
+    if not samplers:
+        raise AttributeError("Fault-tolerance doesn't support an IterableDataset without a sampler as attribute.")
+
+    sampler = [v for v in samplers.values() if type(v) in SUPPORTED_SAMPLERS]
+
+    if not sampler:
+        raise TypeError(f"Fault-tolerance supports only {SUPPORTED_SAMPLERS}.")
+
+    if len(sampler) > 1:
+        raise ValueError(f"A single sampler is supported within an Iterable Dataset. Found {sampler}.")
+
+    if type(sampler[0]) is DistributedSampler and sampler.shuffle:
+        raise ValueError("A `DistributedSampler` sampler shuffle attribute is set to True.")
+
+    if type(sampler[0]) is RandomSampler:
+        raise ValueError("Only SequentialSampler is supported.")
+
+
+def _validate_map_dataset(dataloader: DataLoader):
+    SUPPORTED_SAMPLERS = (RandomSampler, SequentialSampler, DistributedSampler)
+
+    sampler = getattr(dataloader, "sampler", None)
+    if sampler is not None and type(sampler) not in SUPPORTED_SAMPLERS:
+        raise ValueError(f"Fault-tolerance supports only {SUPPORTED_SAMPLERS}.")
+
+    batch_sampler = getattr(dataloader, "batch_sampler", None)
+    if batch_sampler is not None and type(batch_sampler) is not BatchSampler:
+        raise ValueError("Fault-tolerance supports only a BatchSampler.")
+
+    if type(sampler) is DistributedSampler and sampler.shuffle:
+        raise ValueError("A `DistributedSampler` sampler shuffle attribute is set to True.")
+
+    if type(sampler) is RandomSampler:
+        raise ValueError("Only SequentialSampler is supported.")
+
+
+def _validate_fault_tolerant_automatic(dataloader: Iterable, stage: "pl.trainer.states.RunningStage") -> None:
+    """This function is used to validate that Fault-tolerance is possible with the user data."""
+    from pytorch_lightning.trainer.supporters import CombinedLoader, CycleIterator
+
+    if not _FaultTolerantMode.detect_current_mode().is_automatic:
+        return
+
+    if isinstance(dataloader, CombinedLoader):
+        dataloaders = dataloader.loaders
+    else:
+        dataloaders = dataloader
+
+    dl_loaders = []
```

Comment:
I think this kind of extraction of loaders is in other places too. Should we extract it into an utility?

Do you remember where else it's done?

Reply:
Not sure we do anymore as I tried to remove this logic.